### DISCUSSION TITLE
Fatal Python error: PyGILState_Release

### DISCUSSION THREAD

**Discussion by sondrelg** [PyO3/pyo3#2677] [Questions]
Hi! 

When [testing](https://github.com/sondrelg/self-limiters/actions/runs/3224315473/jobs/5275320799) my pyo3 + pyo3-asyncio library built with maturin I've run into the following:

```
Fatal Python error: PyGILState_Release: thread state 0x7f1e840332c0 must be current when releasing
Python runtime state: finalizing (tstate=0x00005575cd335710)

Thread 0x00007f1e8a7c0740 (most recent call first):
  <no Python frame>
/home/runner/work/_temp/e4644e7f-3d00-4cab-9d0c-bcd97e7a8[27](https://github.com/sondrelg/self-limiters/actions/runs/3224315473/jobs/5275320799#step:10:28)0.sh: line 3:  [28](https://github.com/sondrelg/self-limiters/actions/runs/3224315473/jobs/5275320799#step:10:29)64 Aborted                 (core dumped) python bench.py s --count 100 --iterations 30 --target 3.5
```

Googling the relevant parts of the error doesn't really help. Any idea where to start debugging this?

Opened this as a discussion instead of an issue, since I'm not sure it's related to maturin 🤷 I was using `maturin develop` here though. Will try to see if the issue replicates when using a release build.

**Comment by messense** [comment]
I don't think it's related to maturin, it might be related to pyo3 or pyo3-asyncio, I'm going to transfer this to pyo3.

Does it always reproduce or it only occurs sometimes? The failed action you referred to only failed on Python 3.10, did it ever fail on other Python versions?

**Reply by sondrelg** [reply]
Only occurs sometimes. So far 2/5 runs or so. 

I've only tested on 3.10 so far, but will do a bit more testing. 

Never had this happen during local development, only in Github actions 👍

**Reply by messense** [reply]
I just tried locally and also can not reproduce it. Perhaps you can use https://github.com/mxschmitt/action-tmate to get a shell on GitHub Actions to help debugging.

**Reply by sondrelg** [reply]
I was able to reproduce on [Python 3.11](https://github.com/sondrelg/self-limiters/actions/runs/3232727857/jobs/5293690269), but seems like it's happening relatively infrequently. 

Any tips for how you would debug using tmate, if you were able to reproduce? Trying to reproduce now, but not sure what to do if I'm able to 😅

**Reply by messense** [reply]
[Getting a stack trace via the core dump file](https://stackoverflow.com/a/5115653) would be a good start.

**Reply by messense** [reply]
Finally got a coredump locally, see https://gist.github.com/messense/575b65117432e3c7b2ff5af39052b7c2

cc @davidhewitt

**Reply by messense** [reply]
Though that was using pyo3 0.16, I will try to see if I can get one on pyo3 0.17.

**Reply by messense** [reply]
I'm unable to reproduce the issue on pyo3 0.17.2 after many tries, maybe it's already fixed?

**Reply by sondrelg** [reply]
[This run](https://github.com/sondrelg/self-limiters/actions/runs/3233278530/jobs/5294926461) should be using 0.17.2 I believe. I'm flying the next 12 hours, but will try to reproduce more then!

**Reply by messense** [reply]
With a bit of luck I just got one for pyo3 0.17.2 locally: https://gist.github.com/messense/8cb0e20f67a83a2758b96080006a4735

It's a bit different than pyo3 0.16, no more `Fatal Python error: PyGILState_Release: thread state 0x7f1e840332c0 must be current when releasing`, only a segfault.

```
(gdb) up
#3  0x00007fe7ce98364b in pyo3_ffi::object::Py_DECREF (op=0x7fe7cecd2890) at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/pyo3-ffi-0.17.2/src/object.rs:427
427	            _Py_Dealloc(op)
(gdb) p op
$1 = (*mut pyo3_ffi::object::PyObject) 0x7fe7cecd2890
(gdb) p *op
$2 = pyo3_ffi::object::PyObject {ob_refcnt: 0, ob_type: 0x55ab060ddba0 <PyTuple_Type>}
```

**Reply by davidhewitt** [reply]
Hmm, there's not anything which obviously sticks out to me there. The tuple is created by PyO3 using only types which we should have correct implementations for.

It smells a bit like possible heap corruption somewhere else in the program. Can you try running with an address sanitiser and see what happens? E.g. in PyO3 we can run our tests with sanitiser like so: https://github.com/PyO3/pyo3/blob/43261c7b8c7d9c703539662fa5ee47a2eabca0f7/noxfile.py#L224

**Reply by sondrelg** [reply]
Sorry I just remembered this was still open. The package had some lua scripts it would load at runtime, and after inlining those lua scripts this hasn't happened since - not sure if that's helpful 🙂

I'm happy to leave this for now, it might not be related to pyo3 after all
